src/transformStateWithClones.js

```diff
@@ -7,7 +7,25 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const resultState = { ...state };
+  const result = [];
+
+  for (const key of actions) {
+    if (key.type === 'addProperties') {
+      Object.assign(resultState, key.extraData);
+    } else if (key.type === 'removeProperties') {
+      for (const step of key.keysToRemove) {
+        delete resultState[step];
+      }
+    } else if (key.type === 'clear') {
+      for (const keys in resultState) {
+        delete resultState[keys];
+      }
+    }
+    result.push({ ...resultState });
```

Comment:
Pushing a copy of the object to the array should be done at the end of the loop cycle but outside of the 'switch' block, after the current action is processed. This ensures that the same action is not repeated unnecessarily.